cpp/src/jni/orc/jni_wrapper.cpp

```diff
@@ -0,0 +1,287 @@
+    env->ThrowNew(exception_class, error_message.c_str());
+  }
+
+  return ret;
+}
+
+jint JNI_OnLoad(JavaVM* vm, void* reserved) {
+  JNIEnv* env;
+  if (vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION) != JNI_OK) {
+    return JNI_ERR;
+  }
+
+  io_exception_class = CreateGlobalClassReference(env, "Ljava/io/IOException;");
+  exception_class = CreateGlobalClassReference(env, "Ljava/lang/Exception;");
+
+  orc_field_node_class =
+      CreateGlobalClassReference(env, "Lorg/apache/arrow/adapter/orc/OrcFieldNode;");
+  orc_field_node_constructor = GetMethodID(env, orc_field_node_class, "<init>", "(II)V");
+
+  orc_memory_class = CreateGlobalClassReference(
+      env, "Lorg/apache/arrow/adapter/orc/OrcMemoryJniWrapper;");
+  orc_memory_constructor = GetMethodID(env, orc_memory_class, "<init>", "(JJJJ)V");
+
+  record_batch_class =
+      CreateGlobalClassReference(env, "Lorg/apache/arrow/adapter/orc/OrcRecordBatch;");
+  record_batch_constructor =
+      GetMethodID(env, record_batch_class, "<init>",
+                  "(I[Lorg/apache/arrow/adapter/orc/OrcFieldNode;"
+                  "[Lorg/apache/arrow/adapter/orc/OrcMemoryJniWrapper;)V");
+
+  env->ExceptionDescribe();
+
+  return JNI_VERSION;
+}
+
+void JNI_OnUnload(JavaVM* vm, void* reserved) {
+  JNIEnv* env;
+  vm->GetEnv(reinterpret_cast<void**>(&env), JNI_VERSION);
+  env->DeleteGlobalRef(io_exception_class);
+  env->DeleteGlobalRef(exception_class);
+  env->DeleteGlobalRef(orc_field_node_class);
+  env->DeleteGlobalRef(orc_memory_class);
+  env->DeleteGlobalRef(record_batch_class);
+
+  buffer_holder_.Clear();
+  orc_stripe_reader_holder_.Clear();
+  orc_reader_holder_.Clear();
+}
+
+std::shared_ptr<ORCFileReader> GetNativeReader(jlong id) {
+  return orc_reader_holder_.Lookup(id);
+}
+
+std::shared_ptr<RecordBatchReader> GetStripeReader(jlong id) {
+  return orc_stripe_reader_holder_.Lookup(id);
+}
+
+std::string JStringToCString(JNIEnv* env, jstring string) {
+  int32_t jlen, clen;
+  clen = env->GetStringUTFLength(string);
+  jlen = env->GetStringLength(string);
+  std::vector<char> buffer(clen);
+  env->GetStringUTFRegion(string, 0, jlen, buffer.data());
+  return std::string(buffer.data(), clen);
+}
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+JNIEXPORT jlong JNICALL Java_org_apache_arrow_adapter_orc_OrcReaderJniWrapper_open(
+    JNIEnv* env, jobject this_obj, jstring file_path) {
+  std::shared_ptr<arrow::io::ReadableFile> in_file;
+
+  std::string path = JStringToCString(env, file_path);
+
+  arrow::Status ret;
+  if (path.find("hdfs://") == 0) {
+    env->ThrowNew(io_exception_class, "hdfs path not support yet.");
+  } else {
+    ret = arrow::io::ReadableFile::Open(path, &in_file);
+  }
+
+  if (ret.ok()) {
+    std::unique_ptr<ORCFileReader> reader;
+
+    ret = ORCFileReader::Open(
+        std::static_pointer_cast<arrow::io::RandomAccessFile>(in_file),
+        arrow::default_memory_pool(), &reader);
```

Comment:
we should file a JIRA to make this configurable (I can imagine wanting to use whather the java allocator here as well as an implementation of memory pool)

Reply:
As I replied above, I will integrate the memory_pool with java allocator in a separate PR.